<center><h1><b>Trabalho Computacional 01 </b></h1></center>

Autor: Rômulo Bandeira Pimentel Drumond

<h2>2. Questão 02</h2>

Inicialmente o conjunto de dados foi carregado no computador, organizado e separado entre variáveis dependente e independente. O código abaixo foi utilizado para essa função:

In [1]:
import pandas as pd
import numpy as np

labels = ["Density", "Percent body fat", "Age (years)", "Weight (lbs)", "Height (inches)", "Neck circum. (cm)",
          "Chest circum. (cm)", "Abdomen circum. (cm)", "Hip circum. (cm)", "Thigh circum. (cm)", "Knee circum. (cm)",
          "Ankle circum. (cm)", "Biceps circum. (cm)","Forearm circum. (cm)","Wrist circum. (cm)"]
dataset = pd.read_csv('dataset_bodyfat.txt', delim_whitespace=True, names = labels)

y_label = ["Percent body fat"]
y_data = dataset[y_label]

x_label = labels[2:] # excluindo densidade corporal
x_data = dataset[x_label]

/home/romulo/Dropbox/Mestrado/[00 old]/[DISCIPLINA] Inteligência Computacional Aplicada/venv/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)


In [2]:
dataset.head()

Density  Percent body fat  Age (years)  Weight (lbs)  Height (inches)  \
0   1.0708              12.3           23        154.25            67.75   
1   1.0853               6.1           22        173.25            72.25   
2   1.0414              25.3           22        154.00            66.25   
3   1.0751              10.4           26        184.75            72.25   
4   1.0340              28.7           24        184.25            71.25   

   Neck circum. (cm)  Chest circum. (cm)  Abdomen circum. (cm)  \
0               36.2                93.1                  85.2   
1               38.5                93.6                  83.0   
2               34.0                95.8                  87.9   
3               37.4               101.8                  86.4   
4               34.4                97.3                 100.0   

   Hip circum. (cm)  Thigh circum. (cm)  Knee circum. (cm)  \
0              94.5                59.0               37.3   
1              98.7                58.7               37.3   
2              99.2                59.6               38.9   
3             101.2                60.1               37.3   
4             101.9                63.2               42.2   

   Ankle circum. (cm)  Biceps circum. (cm)  Forearm circum. (cm)  \
0                21.9                 32.0                  27.4   
1                23.4                 30.5                  28.9   
2                24.0                 28.8                  25.2   
3                22.8                 32.4                  29.4   
4                24.0                 32.2                  27.7   

   Wrist circum. (cm)  
0                17.1  
1                18.2  
2                16.6  
3                18.2  
4                17.7

<h3><b>
2.1 Formulação do modelo estatístico e estimação de parâmetros
</b></h3>

O modelo adotado foi o linear múltiplo, análogo ao estudado na questão 01, mas em vez de:

$$y=ax+b$$

Tem-se:

$$y=\sum_{i=1}^{p} a_ix_i +b$$

Onde $p$ é o número de parâmetros, $x_i$ é a $i$-ésima variável independente e $b$ o intercepto. Para escrever numa forma vetorial, ou mais compacta, fazemos:

$$x_0=1$$
$$a_0=b$$

$$\vec{x}=[x_0, x_1,\dots,x_p]^t$$
$$\vec{a}=[a_0, a_1,\dots,a_p]^t$$

$$y=\vec{a}^{\hspace{.05cm}t}\vec{x}$$

Considerando um erro aleatório aditivo gaussiano, obtem-se:

$$y=\vec{a}^{\hspace{.05cm}t}\vec{x}+\epsilon$$

Onde $\epsilon$ representa essa variável aleatória.

Para a estimação dos parâmetros, ou seja, estimação do vetor $\vec{a}$, considera-se o número de amostras igual a $n$ e constroem-se as variáveis:

$$\vec{y}_{nx1}=[y_1,\dots,y_n]^t$$

$$
\tilde{X}_{nx(p+1)} =
\begin{bmatrix}
    1 & x_{11} & x_{12} & \dots  & x_{1p} \\
    1 & x_{21} & x_{22} & \dots  & x_{2p} \\
    \vdots & \vdots & \vdots & \ddots & \vdots \\
    1 & x_{n1} & x_{n2} & \dots  & x_{np}
\end{bmatrix}
$$

Onde $y_i$ é o valor de saída da $i$-ésima amostra e $x_{ij}$ é o valor da $j$-ésima variável de entrada da $i$-ésima amostra. Dessa forma reescrevemos a equação original:

$$\vec{y}_{nx1}=\tilde{X}_{nx(p+1)}\vec{a}_{(p+1)x1}+\vec{\epsilon}_{nx1}$$

Onde $\vec{\epsilon}$ é uma variável aletória de dimensão $n$. De forma a minimizar a norma euclidiana do erro temos como solução `\cite{boyd}`:

$$\vec{a}=\tilde{X}^{\dagger} \vec{y}$$

Onde $\tilde{X}^{\dagger}$ representa a pseudo-inversa da matriz $\tilde{X}$.

Como as variáveis possuem escalas diferentes, que podem ser vistas na `\ref{tab:statsDataset}`, faz-se necessário o uso de normalização. A metodologia utilizada foi `\cite{https://en.wikipedia.org/wiki/Normalization_(statistics)}`:

$$u_i=\frac{x_i-\hat{\mu}_i}{\hat{\sigma}_i}$$

Onde $u_i$ é o novo valor da variável $x_i$, $\hat{\mu}_i$ é a estimativa da média e $\hat{\sigma}_i$ a estimativa do desvio-padrão da $i$-ésima variável de entrada.

<b>Tabela: Estatísticas do conjunto de dados.</b>

In [3]:
tab = x_data.describe()
tab = tab.loc[["mean", "std", "min", "max", "25%", "50%", "75%"],:]
tab

Age (years)  Weight (lbs)  Height (inches)  Neck circum. (cm)  \
mean    44.884921    178.924405        70.148810          37.992063   
std     12.602040     29.389160         3.662856           2.430913   
min     22.000000    118.500000        29.500000          31.100000   
max     81.000000    363.150000        77.750000          51.200000   
25%     35.750000    159.000000        68.250000          36.400000   
50%     43.000000    176.500000        70.000000          38.000000   
75%     54.000000    197.000000        72.250000          39.425000   

      Chest circum. (cm)  Abdomen circum. (cm)  Hip circum. (cm)  \
mean          100.824206             92.555952         99.904762   
std             8.430476             10.783077          7.164058   
min            79.300000             69.400000         85.000000   
max           136.200000            148.100000        147.700000   
25%            94.350000             84.575000         95.500000   
50%            99.650000             90.950000         99.300000   
75%           105.375000             99.325000        103.525000   

      Thigh circum. (cm)  Knee circum. (cm)  Ankle circum. (cm)  \
mean           59.405952          38.590476           23.102381   
std             5.249952           2.411805            1.694893   
min            47.200000          33.000000           19.100000   
max            87.300000          49.100000           33.900000   
25%            56.000000          36.975000           22.000000   
50%            59.000000          38.500000           22.800000   
75%            62.350000          39.925000           24.000000   

      Biceps circum. (cm)  Forearm circum. (cm)  Wrist circum. (cm)  
mean            32.273413             28.663889           18.229762  
std              3.021274              2.020691            0.933585  
min             24.800000             21.000000           15.800000  
max             45.000000             34.900000           21.400000  
25%             30.200000             27.300000           17.600000  
50%             32.050000             28.700000           18.300000  
75%             34.325000             30.000000           18.800000

<small><small>Fonte: Elaborada pelo próprio autor.</small></small>

O código abaixo aplica a normalização das variáveis e a regressão linear múltipla:

In [4]:
# Normalização das variáveis de entrada
x_normal = (x_data - x_data.mean()) / x_data.std()
tab2 = x_normal.describe()
tab2 = tab2.loc[["mean", "std", "min", "max", "25%", "50%", "75%"],:]
tab2

Age (years)  Weight (lbs)  Height (inches)  Neck circum. (cm)  \
mean  2.079465e-16 -6.837564e-16     1.476773e-15       1.380554e-14   
std   1.000000e+00  1.000000e+00     1.000000e+00       1.000000e+00   
min  -1.815970e+00 -2.056010e+00    -1.109757e+01      -2.835175e+00   
max   2.865812e+00  6.268488e+00     2.075209e+00       5.433323e+00   
25%  -7.248764e-01 -6.779508e-01    -5.183959e-01      -6.549240e-01   
50%  -1.495727e-01 -8.249316e-02    -4.062664e-02       3.264826e-03   
75%   7.233019e-01  6.150429e-01     5.736482e-01       5.894643e-01   

      Chest circum. (cm)  Abdomen circum. (cm)  Hip circum. (cm)  \
mean       -5.110550e-15          2.992315e-15      6.797913e-15   
std         1.000000e+00          1.000000e+00      1.000000e+00   
min        -2.553143e+00         -2.147435e+00     -2.080492e+00   
max         4.196180e+00          5.151039e+00      6.671532e+00   
25%        -7.679527e-01         -7.401368e-01     -6.148418e-01   
50%        -1.392812e-01         -1.489327e-01     -8.441611e-02   
75%         5.398027e-01          6.277473e-01      5.053335e-01   

      Thigh circum. (cm)  Knee circum. (cm)  Ankle circum. (cm)  \
mean        3.202024e-15      -1.091896e-14       -9.727668e-15   
std         1.000000e+00       1.000000e+00        1.000000e+00   
min        -2.324965e+00      -2.317964e+00       -2.361435e+00   
max         5.313200e+00       4.357535e+00        6.370677e+00   
25%        -6.487588e-01      -6.698205e-01       -6.504131e-01   
50%        -7.732497e-02      -3.751390e-02       -1.784071e-01   
75%         5.607761e-01       5.533300e-01        5.296021e-01   

      Biceps circum. (cm)  Forearm circum. (cm)  Wrist circum. (cm)  
mean        -2.414295e-15          5.286776e-16        3.418782e-16  
std          1.000000e+00          1.000000e+00        1.000000e+00  
min         -2.473597e+00         -3.792707e+00       -2.602615e+00  
max          4.212325e+00          3.086128e+00        3.395768e+00  
25%         -6.862710e-01         -6.749616e-01       -6.745631e-01  
50%         -7.394653e-02          1.787067e-02        7.523482e-02  
75%          6.790471e-01          6.612149e-01        6.108047e-01

In [5]:
# Ordinary Least Squares
from numpy.linalg import pinv

def ols(x, y):
    n = len(y) # tamanho do dataset

    # y = Xa
    ones = np.ones((n, 1))
    X = np.concatenate((ones, x), axis=1)

    # pseudo-inversa de X
    X_cross = pinv(X)

    # a = X_cross*y
    a = np.matmul(X_cross, y)
    return a

a = ols(x_normal.values, y_data.values)
print("O parâmetros ótimos do modelo linear são: \na = {}".format(np.squeeze(a)))

O parâmetros ótimos do modelo linear são: 
a = [19.15079365  0.78231757 -2.59931472 -0.25489971 -1.1439878  -0.2011861
 10.29539552 -1.48683658  1.23951286  0.0368553   0.2949036   0.54867061
  0.91340275 -1.51300424]


In [6]:
# Comparação com scikit-learn
from sklearn import linear_model
reg = linear_model.LinearRegression()
reg.fit (x_normal, y_data)
temp = np.concatenate((reg.intercept_, reg.coef_[0]))
print("a_sklearn = {}".format(temp))

a_sklearn = [19.15079365  0.78231757 -2.59931472 -0.25489971 -1.1439878  -0.2011861
 10.29539552 -1.48683658  1.23951286  0.0368553   0.2949036   0.54867061
  0.91340275 -1.51300424]


<h3><b>
2.2 Análise dos resíduos
</b></h3>

Devido à dificuldade em analisar visualmente um modelo com 13 variáveis de entrada, a metodologia adotada para validar o modelo foi:
  * Cálculo do Coeficiente de Determinação ($R^2$);
  * Análise do Histograma dos resíduos com teste de hipótese de Shapiro-Wilk para normalidade;
  * Análise do Boxplot dos resíduos para identificação de possíveis outliers;
  * Análise de sensibilidade do modelo aos outliers.
  
O código abaixo foi utlizado para gerar o histograma e boxplot dos resíduos assim como calcular o valor de $R^2$ e aplicar o teste de Shapiro-Wilk:

In [7]:
import plotly.offline as plt
import plotly.graph_objs as go
from scipy import stats

plt.init_notebook_mode(connected=True) # habilitando plot dentro do jupyter notebook

# Função que calcula o R2
def r2(y, e_vec):
    y_mean = np.mean(y)
        
    SQ_e = np.dot(np.transpose(e_vec), e_vec)          # sum of squared errors
    S_yy = sum([(y_i - y_mean)**2 for y_i in y]) # sum of errors for naive predictor

    R2 = 1 - SQ_e/S_yy
    return R2

def plotHistogram(e_vec):
    # Histograma dos resíduos
    data = [go.Histogram(
        x=e_vec,
        histnorm="probability density",
    )]

    titlefont=dict(family='Courier New, monospace', size=18, color='#7f7f7f')
    layoutHist = go.Layout(
        title='Histograma dos resíduos',
        xaxis=dict(title='Valor do erro', titlefont=titlefont),
        yaxis=dict(title='Densidade de probabilidade', titlefont=titlefont)
    )

    fig = go.Figure(data=data, layout=layoutHist)
    plt.iplot(fig)

# Boxplot dos resíduos
def plotBoxplot(e_vec):
    data = [
        go.Box(
            y= e_vec,
            boxpoints='all',
            boxmean='sd',
            jitter=0.3,
            pointpos=-1.8
        )
    ]
    
    layoutBoxplot = go.Layout(title="Boxplot dos resíduos")

    fig = go.Figure(data=data, layout=layoutBoxplot)
    plt.iplot(fig)

# criando a matriz X_normal em função do dataframe x_normal
n = len(y_data)
ones = np.ones((n, 1))
X_normal = np.concatenate((ones, x_normal.values), axis=1)

y_pred = np.dot(X_normal, a)
e_vec = np.subtract(y_data.values, y_pred)
e_vec = e_vec.transpose()[0]



plotHistogram(e_vec)
plotBoxplot(e_vec)
R2 = r2(y_data.values, e_vec)
print("R2: {}".format(R2[0]))
print("P-value: {}".format(stats.shapiro(e_vec)[1]))

R2: 0.7490499691482748
P-value: 0.21762634813785553


Duas suposições do modelo se concretizaram:
  * Os resíduos apresentaram média 0 (considerando $10^{-15}$ como praticamente zero);
  * A hipótese de normalidade foi aceita após o teste de Shapiro-Wilk.

A hipótese da presença de outliers foi rejeitada pois no boxplot não foi encontrado uma ou mais amostras com erro discrepante quando comparada com as outras.

O valor de $R^2$ foi relativamente alto, mostrando que o modelo linear teve um bom ajuste nos dados.

<h3><b>
2.3 Análise de sensibilidade do modelo em relação as variáveis preditoras
</b></h3>

De forma simplificada podemos analisar a importância de cada variável preditora pelo módulo do seu coeficiente multiplicativo, visto que as variáveis de entrada foram normalizadas. Para tal o código abaixo gerou a Tabela \ref{tab:modCoef}:

In [8]:
newNames = ['Age', 'Weight', 'Height', 'Neck', 'Chest', 'Abdo.', 'Hip',
            'Thigh', 'Knee', 'Ankle', 'Biceps', 'Forearm', 'Wrist']
var_nomes = np.insert(newNames, 0, "Intercepto")
parametros = pd.DataFrame(columns = ["Coeficiente"], index=var_nomes, data=a) # cria dataframe
display(parametros)

param_abs = parametros.apply(np.absolute)
param_abs.columns = ["Módulo do coeficiente"]
param_abs = param_abs.sort_values("Módulo do coeficiente", ascending=False)
display(param_abs)

Coeficiente
Interce    19.150794
Age         0.782318
Weight     -2.599315
Height     -0.254900
Neck       -1.143988
Chest      -0.201186
Abdo.      10.295396
Hip        -1.486837
Thigh       1.239513
Knee        0.036855
Ankle       0.294904
Biceps      0.548671
Forearm     0.913403
Wrist      -1.513004

Módulo do coeficiente
Interce              19.150794
Abdo.                10.295396
Weight                2.599315
Wrist                 1.513004
Hip                   1.486837
Thigh                 1.239513
Neck                  1.143988
Forearm               0.913403
Age                   0.782318
Biceps                0.548671
Ankle                 0.294904
Height                0.254900
Chest                 0.201186
Knee                  0.036855

De forma complementar o \textit{boxplot} do módulo dos coeficientes foi gerando, podendo ser visto da Figura \ref{fig:boxplotModCoef}.

In [9]:
# Boxplot do módulo dos coeficientes
def boxplotCoef(coef_abs):
    data = [
        go.Box(
            y= coef_abs,
            boxpoints='all',
            boxmean='sd',
            jitter=0.3,
            pointpos=-1.8
        )
    ]
    
    layoutBoxplot = go.Layout(title="Boxplot dos módulos dos coeficientes")

    fig = go.Figure(data=data, layout=layoutBoxplot)
    plt.iplot(fig)
    
boxplotCoef(param_abs.iloc[1:].values.transpose()[0])

Podemos então enumerar as 5 variáveis mais importantes para o modelo:

In [10]:
top5 = param_abs.iloc[1:6]
top5

Módulo do coeficiente
Abdo.               10.295396
Weight               2.599315
Wrist                1.513004
Hip                  1.486837
Thigh                1.239513

O resultado parece ser coerente, visto que o primeiro e segundo lugar são respectivamente a circunferência do abdômem e o peso, que intuitivamente, para autor deste trabalho, são fortemente associados com à porcentagem de gordura corporal. Agora, se quisermos analisar que variáveis são mais importantes para a construção do modelo devemos estabelecer uma metodologia de análise. Para isso foi calculado o $R^2$ e feito o teste de Shapiro-Wilk para cada combinação possível das variáveis de entrada. O código abaixo aplica o teste extensivo e apresenta os resultados na Tabela \ref{tab:analiseDeSensibilidade}:

In [11]:
import itertools

n = x_data.shape[1] # inicialmente 13 variáveis de entrada
opcoes = list(map(list, itertools.product([0, 1], repeat=n))) # todas as combinações possíveis das variáveis

print("Número de possibilidades: {}".format(len(opcoes)))

Número de possibilidades: 8192


In [12]:
data = []
for opcao in opcoes:
    # retirar variáveis que serão utilizadas
    indices = [i for i, x in enumerate(opcao) if x == 1] # salvando índices em que opcao[i]=1
    x = x_normal.iloc[:,indices] # pegando as colunas referentes aos índices em que o opcao[i]=1

    a = ols(x.values, y_data.values) # parâmetros ótimos da regressão

    # calcular e_vec
    N = len(y_data) # quantidade de amostras
    X = np.concatenate((np.ones((N, 1)), x.values), axis=1)
    y_pred = np.dot(X, a)

    e_vec = np.subtract(y_data.values, y_pred).transpose()[0]

    # R2 e Shapiro-Wilk
    R2 = r2(y_data.values, e_vec)
    sw = stats.shapiro(e_vec)[1]

    # salvando informações
    data.append(opcao + [R2[0], sw])

In [13]:
cabecalho = newNames + ["R2", "SW P-value"]
analiseDeSensibilidade = pd.DataFrame(data, columns=cabecalho) # dataframe com os resultados das simulações
analiseDeSensibilidade = analiseDeSensibilidade.sort_values("R2", ascending=False)
analiseDeSensibilidade.head()

Age  Weight  Height  Neck  Chest  Abdo.  Hip  Thigh  Knee  Ankle  \
8191    1       1       1     1      1      1    1      1     1      1   
8175    1       1       1     1      1      1    1      1     0      1   
7935    1       1       1     1      0      1    1      1     1      1   
7919    1       1       1     1      0      1    1      1     0      1   
7167    1       1       0     1      1      1    1      1     1      1   

      Biceps  Forearm  Wrist        R2  SW P-value  
8191       1        1      1  0.749050    0.217626  
8175       1        1      1  0.749046    0.225888  
7935       1        1      1  0.748989    0.220506  
7919       1        1      1  0.748981    0.232735  
7167       1        1      1  0.748496    0.235241

Os cinco melhores resultados, "rankeados" pelo valor de $R^2$, estão contidos na Tabela \ref{tab:analiseDeSensibilidade}. Nessa tabela, na coluna referente a cada variável, um '1' representa a presença da dita variável no modelo e um '0' sua ausência. Conclui-se que para obter o melhor valor de $R^2$ precisa-se de todas variáveis de entrada, no entanto, podemos estudar o *trade-off* que existe entre o número de variáveis e o valor de $R^2$. Para isso formularemos o problema de otimização:

$$\text{minimizar}\hspace{0.5cm} f_o(x)=\big(-R^2(x), x\big)$$
$$\text{sujeito a:} \hspace{0.5cm} 0 \leq x \leq 1 $$

Onde $x$ é o número de variáveis normalizado:

$$x=\frac{\text{número de variáveis utilizadas}}{\text{número total de variáveis}}$$

Perceba que o valor negativo em $R^2$ indica que pretendemos maximizar essa métrica, ou seja, minimizar seu negativo assim como  minimizar $x$, o número de variáveis normalizado.

Foi feito o gráfico da região de possíveis valores de $f_o(x)$, que pode ser visto da Figura \ref{fig:grafOtimo}:

In [14]:
import copy
data_temp = copy.deepcopy(data)

R2_points = []
nVar_points = []

for linha in data_temp:
    R2_points.append(-1*linha[-2])
    nVar_points.append(sum(linha[:-2])/len(linha[:-2]))

In [15]:
traceData = go.Scatter(
    x = R2_points, 
    y = nVar_points, 
    mode='markers',
    name='dataset')
layoutData = go.Layout(
    title = "Gráfico da otimização",
    xaxis=dict(title="-R2"),
    yaxis=dict(title="Número de variáveis normalizado"))

fig = go.Figure(data=[traceData], layout=layoutData)
plt.iplot(fig)

Na Figura \ref{fig:grafOtimo} não existe um ponto que possua simultâneamente o menor valor de $x$ e de $-R^2$, o que existe é uma fronteira de Pareto \cite{boyd}. Para encontrar um único ponto ótimo será necessário escalonar o problema, assumindo o formato:

$$\text{minimizar}\hspace{0.5cm} f_{o_2}(x)=\lambda^t\big(-R^2(x), x\big)=-\lambda_1R^2+\lambda_2x$$
$$\text{sujeito a:} \hspace{0.5cm} 0 \leq x \leq 1 $$

Como $x$ e $R^2$ estão ambos no intervalo $[0,1]$ é coerente estabelecer um $\gamma \in [0,1]$ para representar o *trade-off* entre os objetivos, obtendo-se:

$$\text{minimizar}\hspace{0.5cm} f_{o_2}(x)=-(1-\gamma)R^2+\gamma x$$
$$\text{sujeito a:} \hspace{0.5cm} 0 \leq x \leq 1 $$

Assim para $\gamma=1$ o problema de otimização se torna minimizar apenas o número de variáveis de entrada, resultando num modelo linear contendo apenas o intercepto. Quando $\gamma=0$ o problema se torna maximizar o valor de $R^2$ que, como visto anteriormente, tem como resposta usar todas as variáveis possíveis de entrada.

Como $x$ só pode assumir 8192 valores discretos, podemos avaliar a função $f_{o_2}(x)$ para cada ponto do espaço realizável. Como exemplo adotamos $\gamma=0,5$ e o resultado pode ser visto na Tabela \ref{tab:tabGamma5}.

In [16]:
import copy
import ipywidgets as widgets
from IPython.display import clear_output
from IPython.display import display, HTML

# função objetivo
def fo(R2, nVar, gamma):
    return -R2*(1-gamma) + gamma*nVar # gamma atribui peso, em porcentagem ao número de variáveis

# Widgets que permitem mudar o valor de gamma
gammaTxt = widgets.FloatText(
    value=0.5,
    min=0,
    max=1,
    step=0.05,
    description='gamma:'
)
gammaSlide = widgets.FloatSlider(
    value=0.5,
    min=0,
    max=1,
    step=0.05
)

display(gammaTxt, gammaSlide)
mylink = widgets.jslink((gammaTxt, 'value'), (gammaSlide, 'value'))

cabecalho = newNames + ["R2", "SW P-value", "fo"]

def acharOtimo(change):
    clear_output()
    display(gammaTxt, gammaSlide)
    gamma = change['new'] # novo número de amostras
    
    data_temp = copy.deepcopy(data)

    for linha in data_temp:
        R2 = linha[-2] # de 0 a 1
        nVar = sum(linha[:-2])/len(linha[:-2]) # número de variáveis normalizado
                                               # de 0 a 1
        linha.append(fo(R2, nVar, gamma))

    analiseDeSensibilidade = pd.DataFrame(data_temp, columns=cabecalho) # dataframe com os resultados das simulações
    analiseDeSensibilidade = analiseDeSensibilidade.sort_values("fo", ascending=True)
    display(analiseDeSensibilidade.iloc[0:5])

gammaTxt.observe(acharOtimo, names='value')
gammaSlide.observe(acharOtimo, names='value')

acharOtimo({'new': 0.5})

FloatText(value=0.5, description='gamma:', step=0.05)

FloatSlider(value=0.5, max=1.0, step=0.05)

Age  Weight  Height  Neck  Chest  Abdo.  Hip  Thigh  Knee  Ankle  \
128     0       0       0     0      0      1    0      0     0      0   
2176    0       1       0     0      0      1    0      0     0      0   
129     0       0       0     0      0      1    0      0     0      0   
640     0       0       0     1      0      1    0      0     0      0   
192     0       0       0     0      0      1    1      0     0      0   

      Biceps  Forearm  Wrist        R2  SW P-value        fo  
128        0        0      0  0.661672    0.147161 -0.292375  
2176       0        0      0  0.718798    0.246712 -0.282476  
129        0        0      1  0.702014    0.065204 -0.274084  
640        0        0      0  0.696630    0.594490 -0.271392  
192        0        0      0  0.692859    0.391763 -0.269506

Para valores diferentes de $\gamma$ podemos achar diferentes ótimos, como exemplos tem-se:
  * $\gamma > 0,85$: a solução ótima é não utilizar nenhuma variável de entrada. O modelo linear se reduz ao intercepto, onde intercepto = média da variável de saída;
  * $0,45 \leq \gamma \leq 0,85$: a solução ótima é a que usamos apenas a circunferência do abdômen como variável de entrada;
  * $0,15 \leq \gamma \leq 0,45$: a variável peso aparece junto com a circunferência do abdomen;
  * $\gamma < 0,15$: as outras variáveis de entrada vão aparecendo até que todas fiquem presentes no modelo.

O resultado foi análogo ao anterior, e podemos fazer uma comparação do $\gamma$ com as ordens de grandezas dos coeficientes do modelo completo. No modelo orginal o coeficiente da circunferência do abdômen está uma ordem de grandeza maior que o segundo colocado, o peso, assim percebemos que para uma faixa grande de valores de $\gamma$ o melhor modelo permanece o com uma única variável de entrada. Já o peso, o segundo maior coeficiente no modelo original, tem uma faixa relativamente grande de valores de $\gamma$, visto que seu coeficiente é quase duas vezes o do terceiro lugar, a circunferência do pulso.